ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+rin = string[2:5]
+stri = string[:4]
+ing = string[3:]
+wholeString = string[:]
+# you can also just do
+st = 'string'[:2]
+
+# you will get an empty string if the range is 0 
+empty = "empty"[3:3]
+
+"""
+STRING IS IMMUTABLE
+
+meaning you cannot change any single character inside a string 
+
+Teaversals: ways to loop through the entire string, here are two traverals
+"""
+
+for letter in "string": 
+    print(letter)
+
+x = 0 
+while x < len('string'):
+    print('string'[x])
+    x += 1
+
+"""
+str.format{}
+
+This is a way used to format a string by inputting
+different types into a string, for example
+
+"this is a {} statement since the {}th value is {}".format(True,1,3.4)
+
+you can use e to decide how many decimal places you want the number to be 
+ex. {:.3e} will give three digit after the decimal 
+"""
+# your {} always have to be less than the parameter of format
+print("{}{}".format(1,2,3,4,5))
+# But print("{}{}".format(1)) will give an error 
+
+# L I S T 
+"""
+List 
+
+Lists are sequences of objects, it can be a list of anythinig 
+
+when you write the type list, you have to CAPITALIZE the L
+
+so List[Any] is correct, list[any] is not 
+
+you can traversing the list the same way you traversing a string 
+"""
+
+L = [1,2,3,4]
+for elements in L:
+    print(elements)
+
+#or you can use in range, in this case, i is the INDEX of each element
+for i in range (len(L)):
+    print(L[i])
+
+x = 0
+while x < len(L):
+    print(L[x])
+    x += 1
+
+"""
+YOU CANNOT USE OPERTAOR ON LIST TO MUTATE A LIST
+
+you have to use list.append(smth)
+
+list.extend(smth) takes in a list and add to the list being mutated
+
+list.copy() can create a copy of the list 
+
+list .pop(index) can pop out the speicific element of that index in the list
+
+list.pop() just pops out the last 
+
+list(string) makes a list to all the characters in a string
+
+string.split() makes a list of all the words in a string seperated by empty space
+
+string.split(string) make a list of all the characters, seperated by that string
+
+list.joint(string) make a string out of the list, using the parameter string as a joint
+"""
+
+a = [2,3,4,5]
+a.append(6)
+
+#a will be [2,3,4,5,6]
+
+# use + on list 
+b = [6]
+c = a + b 
+
+a.extend(b)
+#c = [2,3,4,5,6]
+
+# use * on list 
+
+c = b*4
+#c = [6,6,6,6]
+
+c = a.copy()
+# c will become a shallow copy of a 
+"""
+List can be sliced just like string
+"""
+
+slice = ['a',2,'b',3]
+a2 = slice[:2]
+
+"""
+sort() and sorted()
+
+list.sort(key,reverse)
+
+You can write your own function as key, these are called helper functions
+"""
+#This is how you mutate a list, key is a function
+[2,3,4,5].sort(key = len, reverse = False)
+
+#This is how you create a new list from the mutated list
+list = sorted([2,3,4,5],key = len, reverse = False)
+
+"""
+Tuple 
+
+Tuples are just lists that are immutable, tuple uses () instead of []
+
+tuple(list) can turn a list in to a tuple 
+
+"""
+tuple = ('one',)
+string = ('one')
+
+"""
+Dictionary 
```

Comment:
Just a heads up, a Dictionary in Python is the equivalent of what is called a "hash table" or "has map" in other languages. The popularity of these terms is very prevalent so unless you are dealing with someone who works purely in python, you may occasionally see something like "use a hash map" even though "hash map" isn't an actual type defined in Python.